File: ts-src/data/PlayerTypes.ts
import { Player } from "../objects/Player"

export interface IPlayerType {
  frames: number
  img: string
  multiplier?: number
  title: string
  type: string
  value: number
  fallSpeed?: number
  moveSpeed?: number
  flying?: boolean
  movePattern?: string
}

export const playerTypes: IPlayerType[] = [
  {
    frames: 18,
    img: "egg-sprite-blue.png",
    multiplier: 5,
    title: "It is of course the blue egg",
    type: "blue-egg",
    value: 3
  },
  {
    frames: 18,
    img: "egg-sprite.png",
    multiplier: 1,
    title: "It is of course the egg",
    type: "egg",
    value: 1
  },
  {
    frames: 18,
    img: "egg-sprite-red.png",
    multiplier: 2,
    title: "It is of course the red egg",
    type: "red-egg",
    value: 2
  },
  {
    fallSpeed: 20,
    frames: 1,
    img: "silver-egg.png",
    moveSpeed: 0,
    multiplier: 10,
    title: "It is of course the silver egg",
    type: "silver-egg",
    value: 0
  },
  {
    frames: 18,
    img: "egg-sprite-yellow.png",
    multiplier: 10,
    title: "It is of course the yellow egg",
    type: "yellow-egg",
    value: 4
  },
  {
    frames: 18,
    img: "egg-rainbow.png",
    multiplier: 1,
    title: "It goes without saying that this is the rainbow egg",
    type: "rainbow-egg",
    value: 1
  },
  {
    frames: 18,
    img: "blade-sprite.png",
    title: "It is the mean spirited blade",
    type: "blade",
    value: 0,
    flying: true
  },
  {
    frames: 18,
    img: "find-blade-sprite.png",
    title: "It is the mean spirited blade",
    type: "find-blade",
    value: 0,
    movePattern: "seek-egg",
    flying: true
  }
];

File: rust/src/data/player_types.rs
use once_cell::sync::Lazy;
use std::collections::HashMap;

use crate::types::player_type::{PlayerKind, PlayerType};
use crate::types::resource_url::ResourceUrl;

pub static PLAYER_TYPES: Lazy<HashMap<PlayerKind, PlayerType>> = Lazy::new(|| {
    let mut m = HashMap::new();

    m.insert(
        PlayerKind::Egg,
        PlayerType {
            frames: 18,
            img: ResourceUrl::SpriteResource("egg-sprite".into()),
            multiplier: 1,
            title: "It is of course the egg".into(),
            type_: PlayerKind::Egg,
            ..PlayerType::default_player_type()
        },
    );

    m.insert(
        PlayerKind::RedEgg,
        PlayerType {
            frames: 18,
            img: ResourceUrl::SpriteResource("egg-sprite-red".into()),
            multiplier: 2,
            title: "It is of course the red egg".into(),
            type_: PlayerKind::RedEgg,
            ..PlayerType::default_player_type()
        },
    );

    m.insert(
        PlayerKind::BlueEgg,
        PlayerType {
            frames: 18,
            img: ResourceUrl::SpriteResource("egg-sprite-blue".into()),
            multiplier: 5,
            title: "It is of course the blue egg".into(),
            type_: PlayerKind::BlueEgg,
            ..PlayerType::default_player_type()
        },
    );

    m.insert(
        PlayerKind::YellowEgg,
        PlayerType {
            frames: 18,
            img: ResourceUrl::SpriteResource("egg-sprite-yellow".into()),
            multiplier: 10,
            title: "It is of course the yellow egg".into(),
            type_: PlayerKind::YellowEgg,
            ..PlayerType::default_player_type()
        },
    );

    m.insert(
        PlayerKind::RainbowEgg,
        PlayerType {
            frames: 18,
            img: ResourceUrl::SpriteResource("egg-rainbow".into()),
            multiplier: 1,
            title: "It goes without saying that this is the rainbow egg".into(),
            type_: PlayerKind::RainbowEgg,
            ..PlayerType::default_player_type()
        },
    );

    m.insert(
        PlayerKind::SilverEgg,
        PlayerType {
            fall_speed: 20,
            frames: 1,
            img: ResourceUrl::SpriteResource("silver-egg".into()),
            move_speed: 0,
            multiplier: 10,
            title: "It is of course the silver egg".into(),
            type_: PlayerKind::SilverEgg,
            ..PlayerType::default_player_type()
        },
    );

    m.insert(
        PlayerKind::Blade,
        PlayerType {
            frames: 18,
            img: ResourceUrl::SpriteResource("blade-sprite".into()),
            title: "It is the mean spirited blade".into(),
            type_: PlayerKind::Blade,
            flying: true,
            ..PlayerType::default_player_type()
        },
    );

    m.insert(
        PlayerKind::FindBlade,
        PlayerType {
            frames: 18,
            img: ResourceUrl::SpriteResource("find-blade-sprite".into()),
            title: "It is the mean spirited blade".into(),
            type_: PlayerKind::FindBlade,
            move_pattern: "seek-egg".into(),
            flying: true,
            ..PlayerType::default_player_type()
        },
    );

    m
});

pub fn get_player_type(kind: &PlayerKind) -> PlayerType {
    PLAYER_TYPES
        .get(kind)
        .cloned()
        .unwrap_or_else(PlayerType::default_player_type)
}

pub fn sprite_resources() -> Vec<ResourceUrl> {
    PLAYER_TYPES.values().map(|pt| pt.img.clone()).collect()
}
